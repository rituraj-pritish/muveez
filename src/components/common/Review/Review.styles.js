import styled from 'styled-components'
import { space } from 'styled-system'
import theme from 'theme'

export const Wrapper = styled.div`
  background: ${({ theme }) => theme.colors.paper};
  border-radius: ${theme.borderRadius};
  display: flex;
  flex-direction: column;
  height: ${({ fullView }) => fullView ? '100%' : 'fit-content'};
  width: 100%;
  padding: ${theme.spacing()};
  padding-top: ${({ isLoading }) => isLoading ? theme.spacing() : 0};
  padding-bottom: ${theme.spacing(1.5)};
  position: relative;
  overflow: auto;

  // top section
  & > div:first-child {
    background: ${({ theme }) => theme.colors.paper};
    position: sticky;
    top: 0;
  }

  ${space}
`
