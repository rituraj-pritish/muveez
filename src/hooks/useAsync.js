import { useState, useEffect } from 'react'

const STATUS = {
	IDLE: 'idle',
	REQUESTING: 'requesting',
	SUCCESS: 'success',
	ERROR: 'error'
}

export default (func, imediate = false) => {
	const [status, setStatus] = useState(STATUS.IDLE)
	const [error, setError] = useState(null)
	const [data, setData] = useState(null)

	const callFunction = async (payload) => {
		setStatus(STATUS.REQUESTING)
		try {
			const res = await func(payload)
			setStatus(STATUS.SUCCESS)
			setData(res)
		} catch (err) {
			setStatus(STATUS.ERROR)
			setError(err)
		}
	}

	useEffect(() => {
		if(imediate) {
			callFunction()
		}
	}, [])

	return {
		callFunction,
		status,
		error,
		data,
		STATUS,
		requesting: status === STATUS.REQUESTING
	}
}