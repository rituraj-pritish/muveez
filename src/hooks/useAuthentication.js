import { createState, useState } from '@hookstate/core'

import { 
	getUser, 
	login as loginEndpoint, 
	logout as logoutEndpoint 
} from 'api/endpoints/authentication' 

const INITIAL_STATE = {
	isLoading: true,
	isAuthenticated: false,
	user: null
}
const AUTH_STATE = createState(INITIAL_STATE)

export const SESSION_ID_KEY = 'session_id'

export default () => {
	const authState = useState(AUTH_STATE)

	const login = async data => {
		try {
			const res = await loginEndpoint(data)
			if(res.success) {
				localStorage.setItem(SESSION_ID_KEY, res.session_id)
				const user = await getUser()

				authState.set({
					isLoading: false,
					isAuthenticated: true,
					user
				})
			}
		} catch (err) {
			// todo show message
			authState.set({ ...INITIAL_STATE, isLoading: false })
		}
	}

	const checkIfAuthenticated = async () => {
		try {
			const sessionId = localStorage.getItem(SESSION_ID_KEY)
			if(!sessionId) throw new Error('no session id')

			const res = await getUser()
			authState.set({
				isLoading: false,
				isAuthenticated: true,
				user: res,
			})
		} catch (err) {
			authState.set({ ...INITIAL_STATE, isLoading: false })
		}
	}

	const logout = async () => {
		try {
			await logoutEndpoint()
			authState.set({
				...INITIAL_STATE, isLoading: false
			})
			localStorage.removeItem(SESSION_ID_KEY)
			//todo handle error
		// eslint-disable-next-line no-empty
		} catch (err) {}
	}

	return {
		login,
		checkIfAuthenticated,
		...authState.get(),
		logout
	}
}