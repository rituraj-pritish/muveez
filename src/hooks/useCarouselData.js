import api from 'api'
import { useQuery } from 'react-query'

const getQueryKey = path => {
	const key = path.split('/')
	key.shift()
	return key
}

const options = {
	enabled: false
}

export default (paths) => {
	return paths
		.map(path => useQuery(getQueryKey(path), () => api(path), options))
}