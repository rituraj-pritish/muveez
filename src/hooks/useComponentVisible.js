import { useState, useEffect, useRef } from 'react'

export default (initialIsVisible) => {
	const [isComponentVisible, setIsComponentVisible] = useState(initialIsVisible)
	const ref = useRef(null)

	const handleClick = event => {
		if (ref.current && !ref.current.contains(event.target)) {
			setIsComponentVisible(false)
			document.removeEventListener('click', handleClick, true)
		}
	}

	useEffect(() => {
		if (isComponentVisible) {
			document.addEventListener('click', handleClick, true)
		}
	}, [isComponentVisible])

	return [ref, isComponentVisible, setIsComponentVisible]
}
