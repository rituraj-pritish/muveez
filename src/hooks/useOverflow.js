import { useLayoutEffect, useRef, useState } from 'react'

export default () => {
	const ref = useRef()
	const [isOverflowingHeight, setIsOverflowingHeight] = useState(false)
	const [isOverflowingWidth, setIsOverflowingWidth] = useState(false)

	useLayoutEffect(() => {
		if(ref.current) {
			setIsOverflowingHeight(ref.current.scrollHeight > ref.current.offsetHeight)
			setIsOverflowingWidth(ref.current.scrollWidth > ref.current.offsetWidth )
		}
	}, [ref])

	return [ref, isOverflowingHeight, isOverflowingWidth]
}