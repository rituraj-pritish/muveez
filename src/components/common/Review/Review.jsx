import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import Skeleton from '../ui/Skeleton'
import FlexBox from '../ui/FlexBox'
import Text from '../ui/Text'
import Link from '../ui/Link'
import Modal from '../Modal'
import Avatar from '../Avatar'
import { Wrapper } from './Review.styles'
import useOverflow from 'hooks/useOverflow'

const Review = ({ fullView, ...review }) => {
	const { 
		id, 
		content, 
		created_at, 
		author_details 
	} = review
	const [contentRef, isOverflowingHeight] = useOverflow()

	if (!id) return (
		<Wrapper fullView={fullView}>
			<FlexBox
				alignItems='center'
				mb={3}
			>
				<Skeleton
					circle
					width={40}
					height={40}
				/>
				<Skeleton
					ml={3}
					width={150}
					height={20}
				/>
			</FlexBox>
			<Skeleton
				count={!fullView ? 3 : 6}
				my={1}
			/>
		</Wrapper>
	)

	return (
		<Wrapper
			fullView={fullView}
			my={2}
		>
			<FlexBox
				alignItems='center'
				mb={3}
			>
				<Avatar
					{...author_details}
					size={40}
				/>
				<Text
					mx={3}
					bold
					size={2}
				>
					{author_details.name || author_details.username}
				</Text>
				<Text
					color='textTertiary'
					size={2}
				>
					{moment(created_at).format('DD MMM YYYY hh:mm A')}
				</Text>
			</FlexBox>
			<Text
				elRef={contentRef}
				color='textSecondary'
				maxLines={!fullView ? 5 : undefined}
			>
				{content}
			</Text>

			{!fullView && isOverflowingHeight && (
				<Modal
					trigger={
						<Link
							align='right'
							mt={2}
							size={1}
						>
              View full review
						</Link>
					}
					styles={{
						width: '50%',
						height: '50%'
					}}
				>
					<Review
						fullView
						{...review}
					/>
				</Modal>
			)}
		</Wrapper>
	)
}

Review.propTypes = {
	fullView: PropTypes.bool
}

export default Review
