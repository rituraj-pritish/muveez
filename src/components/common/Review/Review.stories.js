import React from 'react'
import Review from '.'
import FlexBox from '../ui/FlexBox'

export default { title: 'Review' }

const DATA = {
	'author': 'Cat Ellington',
	'author_details': {
		'name': 'Cat Ellington',
		'username': 'CatEllington',
		'avatar_path': '/uCmwgSbJAcHqNwSvQvTv2dB95tx.jpg',
		'rating': null
	},
	'content': '(As I\'m writing this review, Darth Vader\'s theme music begins to build in my mind...)\r\n\r\nWell, it actually has a title, what the Darth Vader theme. And that title is "The Imperial March", composed by the great John Williams, whom, as many of you may already know, also composed the theme music for "Jaws" - that legendary score simply titled, "Main Title (Theme From Jaws)".\r\n\r\nNow, with that lil\' bit of trivia aside, let us procede with the fabled film currently under review: Star Wars. It had been at a drive-in theater in some small Illinois town or other where my mother, my older brother, and I had spent our weekly "Movie Date Night" watching this George Lucas directed cult masterpiece from our car in the parking lot. On the huge outdoor screen, the film appeared to be a silent one, but thanks to an old wire-attached speaker, we were able to hear both the character dialogue and soundtrack loud and clear. We even had ourselves a carful of vittles and snacks - walked back to our vehicle, of course, from the wide-opened cinema\'s briefly distant concession stand. Indeed, it had been a lovely summer evening that July.\r\n\r\nFrom the time the film started, with my brother and I following along as our mother sped-read the opening crawl, I began to feel rather antsy, thinking that this movie, the first in a franchise that would soon be world-renowned, was going to be boring, due to its genre being Science Fiction: A respectably likable, but not a passionately lovable genre of mine DURING THAT TIME. I just didn\'t believe I was going to like Star Wars all that much ... But I soon found myself intrigued ... And awed.\r\n\r\nGeorge Lucas is a man with a phenomenal, and I do mean phenomenal imagination. Apart from his human characters (Han, Luke, Leia, and Obi-Wan Kenobi, among others), the droids: C-3P0, R2-D2, R2-series, and IG-88, not to mention those unusual characters like Jabba the Hutt, Yoda, and Chewbacca, just to name a few, are all creations of Lucas\'s phenomenal imagination. And I was completely in awe of each one of these strange beings. Then there was Vader ... And the evil Emperor ... And the Stormtroopers ... And the Spacecraft ... And the galaxies (I\'ll admit that I am a huge lover of the Universe in all its Celestial glory) ... And the magnificent planets ... The Lightsabers ... And so on. Star Wars is a gorgeously shot space opera; it is truly an epic masterpiece. We enjoyed this film tremendously. And my brother was a die-hard fan from that night onward. He, my brother, had even received for Christmas that year, nearly every Star Wars action figure that my mother could find, including two of the spacecraft: The Millennium Falcon and Star Destroyer. The Death Star space station had too been wrapped beneath our Christmas tree - tagged with his name. It was totally crazy, what the new Star Wars era. Frenzied! But it was great ... Even still, to this day.\r\n\r\nI don\'t personally know anyone whom has yet to see Star Wars, but that certainly doesn\'t suggest there are still a few people out there who haven\'t. And if you\'re one of the latter, then you should know that this classic space opera comes highly recommended. The entire series is told backwards, so you\'ll definitely want to see Star Wars first, followed by its two sequels: The Empire Strikes Back and Return of the Jedi ... In that order. I trust that you\'ll too discover yourself to be a lifelong cult fan in the wake. ðŸ˜Š',
	'created_at': '2017-02-13T22:23:01.268Z',
	'id': '58a231c5925141179e000674',
	'updated_at': '2017-02-13T23:16:19.538Z',
	'url': 'https://www.themoviedb.org/review/58a231c5925141179e000674'
}

export const review = () => (
	<>
		<Review {...DATA}/>
		<FlexBox my={3}/>
		<Review />
	</>
)