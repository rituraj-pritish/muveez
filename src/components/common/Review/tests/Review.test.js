import React from 'react'
import * as hooks from 'hooks/useOverflow'

// jest.mock('hooks/useOverflow.js', () => () => [jest.fn(), true])

import { render, screen } from 'tests/utils'
import Review from '../index'
import userEvent from '@testing-library/user-event'

const REVIEW = {
	'author': 'TheDarkKnight31',
	'author_details': {
		'name': '',
		'username': 'TheDarkKnight31',
		'avatar_path': null,
		'rating': 10
	},
	'content': 'I will be short. You should understand how hard to make movies based on such a legendary universe (expectation is too high!), with a lot of characters that need screen time, and with limited budget, PLUS in a pandemic situation, - the director made a great job.\r\n\r\nThis is the best adaptation of such a legendary title.',
	'created_at': '2021-04-12T05:39:44.021Z',
	'id': '6073dd204d0e8d0043061b63',
	'updated_at': '2021-04-27T22:52:50.651Z',
	'url': 'https://www.themoviedb.org/review/6073dd204d0e8d0043061b63'
}

describe('Review', () => {
	test('renders review component without full review', async () => {
		jest.spyOn(hooks, 'default').mockImplementation(() => [jest.fn(), false])

		const { container } = render(<Review {...REVIEW}/>)
		expect(screen.getByText('TheDarkKnight31')).toBeInTheDocument()
		expect(screen.getByText('12 Apr 2021 11:09 AM')).toBeInTheDocument()
		expect(screen.queryByText('View full review')).not.toBeInTheDocument()

		expect(container).toMatchSnapshot()
	})

	test('renders review component with full review', async () => {
		jest.spyOn(hooks, 'default').mockImplementation(() => [jest.fn(), true])

		const { container } = render(<Review {...REVIEW}/>)
		expect(screen.getByText('TheDarkKnight31')).toBeInTheDocument()
		expect(screen.getByText('12 Apr 2021 11:09 AM')).toBeInTheDocument()
		expect(screen.getByText('View full review')).toBeInTheDocument()

		userEvent.click(screen.getByText('View full review'))

		expect(screen.getAllByText('TheDarkKnight31')).toHaveLength(2)
		expect(screen.getAllByText('12 Apr 2021 11:09 AM')).toHaveLength(2)
		expect(screen.getAllByText('View full review')).toHaveLength(1)

		expect(container).toMatchSnapshot()
	})
})