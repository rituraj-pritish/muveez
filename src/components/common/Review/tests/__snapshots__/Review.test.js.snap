// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review renders review component with full review 1`] = `
<div>
  <div
    class="css-ahmskn-SkeletonTheme"
  >
    <div
      class="sc-cNKpQo dBMQXU"
    >
      <div
        class="sc-dkPtyc tojMa"
      >
        <div
          class="sc-egiSv bEDQSa"
          height="40"
          width="40"
        >
          <div
            class="sc-bqiQRQ jJORTs"
            height="40"
            width="40"
          />
        </div>
        <div
          class="sc-hKwCoD dAmahb"
          color="textSecondary"
          font-size="2"
          font-weight="bold"
        >
          TheDarkKnight31
        </div>
        <div
          class="sc-hKwCoD jllRuy"
          color="textTertiary"
          font-size="2"
        >
          12 Apr 2021 11:09 AM
        </div>
      </div>
      <div
        class="sc-hKwCoD excitZ"
        color="textSecondary"
        font-size="2"
      >
        I will be short. You should understand how hard to make movies based on such a legendary universe (expectation is too high!), with a lot of characters that need screen time, and with limited budget, PLUS in a pandemic situation, - the director made a great job.

This is the best adaptation of such a legendary title.
      </div>
      <div>
        <div
          class="sc-kfPtPH jkzmwL"
        >
          <div
            class="sc-hKwCoD fqsjRy"
            color="secondary"
            font-size="1"
          >
            <a
              class="sc-eCImvq iuTZXO"
              href=""
            >
              View full review
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Review renders review component without full review 1`] = `
<div>
  <div
    class="css-ahmskn-SkeletonTheme"
  >
    <div
      class="sc-cNKpQo dBMQXU"
    >
      <div
        class="sc-dkPtyc tojMa"
      >
        <div
          class="sc-egiSv bEDQSa"
          height="40"
          width="40"
        >
          <div
            class="sc-bqiQRQ jJORTs"
            height="40"
            width="40"
          />
        </div>
        <div
          class="sc-hKwCoD dAmahb"
          color="textSecondary"
          font-size="2"
          font-weight="bold"
        >
          TheDarkKnight31
        </div>
        <div
          class="sc-hKwCoD jllRuy"
          color="textTertiary"
          font-size="2"
        >
          12 Apr 2021 11:09 AM
        </div>
      </div>
      <div
        class="sc-hKwCoD excitZ"
        color="textSecondary"
        font-size="2"
      >
        I will be short. You should understand how hard to make movies based on such a legendary universe (expectation is too high!), with a lot of characters that need screen time, and with limited budget, PLUS in a pandemic situation, - the director made a great job.

This is the best adaptation of such a legendary title.
      </div>
    </div>
  </div>
</div>
`;
